### PR TITLE
feat(deps): upgrade `expr` from 1.16.0 to 1.16.9 to support `concat` function. Fixes #13175

### LINKED ISSUE #13175: Expr lang in argo: can't use `concat` function
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

Using most functions works as expected, however `concat` results in the expression not being evaluated.

### Version

v3.5.7

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: test-expression-
  namespace: argo
spec:
  entrypoint: main
  arguments:
    parameters:
      - name: expr
        value: "{{= concat(['a', 'b'], ['c', 'd']) | join('\\n') }}"
  templates:
    - name: main
      inputs:
        parameters:
          - name: expr
      script:
        image: alpine:3.6
        command: ["sh"]
        source: |
          echo result: '{{ inputs.parameters.expr }}'
```

### Logs from the workflow controller

```text
time="2024-06-12T16:48:01.582Z" level=info msg="Processing workflow" Phase= ResourceVersion=959144150 namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.590Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.590Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.590Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.591Z" level=info msg="was unable to obtain node for , letting display name to be nodeName" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.591Z" level=info msg="Pod node test-expression-rrl4p initialized Pending" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.648Z" level=info msg="Created pod: test-expression-rrl4p (test-expression-rrl4p)" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.648Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.648Z" level=info msg=reconcileAgentPod namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:01.663Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=959144152 workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.649Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=959144152 namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.650Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.650Z" level=info msg="task-result changed" namespace=argo nodeID=test-expression-rrl4p workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.650Z" level=info msg="node changed" namespace=argo new.message= new.phase=Running new.progress=0/1 nodeID=test-expression-rrl4p old.message= old.phase=Pending old.progress=0/1 workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.651Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.651Z" level=info msg=reconcileAgentPod namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:11.656Z" level=info msg="cleaning up pod" action=terminateContainers key=argo/test-expression-rrl4p/terminateContainers
time="2024-06-12T16:48:11.661Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=959144263 workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.589Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=959144263 namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="task-result changed" namespace=argo nodeID=test-expression-rrl4p workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=test-expression-rrl4p old.message= old.phase=Running old.progress=0/1 workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg=reconcileAgentPod namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="Updated phase Running -> Succeeded" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="Marking workflow completed" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.590Z" level=info msg="Marking workflow as pending archiving" namespace=argo workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.596Z" level=info msg="cleaning up pod" action=deletePod key=argo/test-expression-rrl4p-1340600742-agent/deletePod
time="2024-06-12T16:48:22.603Z" level=info msg="Workflow update successful" namespace=argo phase=Succeeded resourceVersion=959144369 workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.615Z" level=info msg="archiving workflow" namespace=argo uid=bcdc71c8-cee4-4e8b-9d87-780b7c436426 workflow=test-expression-rrl4p
time="2024-06-12T16:48:22.642Z" level=info msg="Queueing Succeeded workflow argo/test-expression-rrl4p for delete in 10m0s due to TTL"
time="2024-06-12T16:48:41.656Z" level=info msg="cleaning up pod" action=killContainers key=argo/test-expression-rrl4p/killContainers
```

### Logs from in your workflow's wait container

```text
time="2024-06-12T16:48:04.442Z" level=info msg="Starting deadline monitor"
time="2024-06-12T16:48:10.447Z" level=info msg="Main container completed" error="<nil>"
time="2024-06-12T16:48:10.447Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2024-06-12T16:48:10.447Z" level=info msg="No output parameters"
time="2024-06-12T16:48:10.447Z" level=info msg="No output artifacts"
time="2024-06-12T16:48:10.470Z" level=info msg="GCS Save path: /tmp/argo/outputs/logs/main.log, key: test-expression-rrl4p/test-expression-rrl4p/main.log"
time="2024-06-12T16:48:11.047Z" level=info msg="Save artifact" artifactName=main-logs duration=576.378086ms error="<nil>" key=test-expression-rrl4p/test-expression-rrl4p/main.log
time="2024-06-12T16:48:11.047Z" level=info msg="not deleting local artifact" localArtPath=/tmp/argo/outputs/logs/main.log
time="2024-06-12T16:48:11.047Z" level=info msg="Successfully saved file: /tmp/argo/outputs/logs/main.log"
time="2024-06-12T16:48:11.067Z" level=info msg="Alloc=26461 TotalAlloc=32942 Sys=39525 NumGC=5 Goroutines=11"
```

### PR DESCRIPTION
**Pull request by jiachengxu**
Fixes #13175 
### Verification
Before this PR, creating the following workflow in the original issue:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: test-expression-
  namespace: argo
spec:
  entrypoint: main
  arguments:
    parameters:
      - name: expr
        value: "{{= concat(['a', 'b'], ['c', 'd']) | join('\\n') }}"
  templates:
    - name: main
      inputs:
        parameters:
          - name: expr
      script:
        image: alpine:3.6
        command: ["sh"]
        source: |
          echo result: '{{ inputs.parameters.expr }}'
```
it logs in the main container:
```
result: {{= concat([a, b], [c, d]) | join(n) }}
```
With this PR, it logs:
```
result: a
b
c
d
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,7 +19,7 @@ require (
 	github.com/coreos/go-oidc/v3 v3.9.0
 	github.com/doublerebel/bellows v0.0.0-20160303004610-f177d92a03d3
 	github.com/evanphx/json-patch v5.8.0+incompatible
-	github.com/expr-lang/expr v1.16.0
+	github.com/expr-lang/expr v1.16.1
```
Comment (agilgur5) [inline go.mod:22]:
```suggestion
	github.com/expr-lang/expr v1.16.9
```

Should probably just bump to latest patch for extra bug fixes etc

Comment (jiachengxu) [reply go.mod]: Good point, just updated the PR to upgrade it to `v1.16.9`

**Comment by agilgur5** [comment]
Looks like some of the tests are failing ([unit](https://github.com/argoproj/argo-workflows/actions/runs/9536721889/job/26283983463?pr=13194) and E2E) as the error message format is ever so slightly different (looks like expr isn't giving a line number back anymore?)

**Comment by jiachengxu** [comment]
> https://github.com/argoproj/argo-workflows/pull/13194#issuecomment-2171695735

Yes, right. Checking the debugging the test locally and it seems like the line number, e.g., `(1:1)` is not there anymore

**Comment by jiachengxu** [comment]
Is it OK if we just update the error message to make it match what `expr` currently returns?

**Comment by agilgur5** [comment]
Yes I think that's fine. This actually might be a bug in `expr`, as it doesn't seem to be mentioned in the release notes anywhere

**Comment by antonmedv** [comment]
Will take a look. Is there an example what reproduces the error?

**Comment by jiachengxu** [comment]
> Will take a look. Is there an example what reproduces the error?

@antonmedv Thanks for taking a look at this. You can use the following program to reproduce the issue:
```golang
package main

import (
	"log"

	"github.com/expr-lang/expr"
)

func main() {
	input := ""
	env := map[string]interface{}{
		"key": "value",
	}
	_, err := expr.Compile(input, expr.Env(env))
	if err != nil {
		log.Fatal(err)
	}
}
```
I also did some experiments and I narrowed the issue down to `v1.16.8`, before `v1.16.8`, the error is:
```
unexpected token EOF (1:1)
```
For version >= `v1.16.8`, the line number in the error is not available anymore, and the error message becomes the following:
```
unexpected token EOF
```

**Comment by antonmedv** [comment]
I took a look on why this error is here. Looks like in this case, as input is empty `""`, Expr does not add any more location information (I don't think it is useful in the case anyway). 

I can just append this string to the error message `+ " (1, 1)"` or we can the test (remote location part, or use `Contains(t, "unexpected token EOF", err.Error())`).

What do you think?

**Comment by jiachengxu** [comment]
> I took a look on why this error is here. Looks like in this case, as input is empty `""`, Expr does not add any more location information (I don't think it is useful in the case anyway).
> 
> I can just append this string to the error message `+ " (1, 1)"` or we can the test (remote location part, or use `Contains(t, "unexpected token EOF", err.Error())`).
> 
> What do you think?

@antonmedv thank you for taking a look at this. Yes, I agree with you that in this case the location info doesn't help that much, I think we can just update our test by removing the location from the expected error message. @agilgur5 Please let me know your thought on this, if you agree, I can update tests this PR.

**Comment by agilgur5** [comment]
I would say yes to both actually 😅

I don't think tests, in general, should rely on _exact_ errors, as error specifics can change (and improving an error message, for instance, is usually not considered a breaking change). So @jiachengxu I think that change makes sense either way.

But I also think `expr` should have consistent errors, it's not clear to me why this behavior changed. It is empty so the line number isn't very useful, but it makes sense to have it for consistency. I also imagine there might be other users who's tests might break, so I think it would be good to add that back as well @antonmedv. Unless you had a good reason to remove it, but it seems like it was unintentional?

**Comment by antonmedv** [comment]
Expr provides errors during compilation from a few sources, some comes with a "snippet" `|----------^`.

My idea was to make this consistent across all errors. If an error contains a snippet `|----^`, add `(line, column)` to the error.

**Comment by antonmedv** [comment]
Also I was refactoring error location information from line and column for offsets start and end. This way it is possible to select the whole token with error.

**Comment by tooptoop4** [comment]
i wonder if this helps with https://github.com/argoproj/argo-workflows/issues/13061 do u happen to use SSO RBAC @jiachengxu  ?

**Comment by jiachengxu** [comment]
> i wonder if this helps with #13061 do u happen to use SSO RBAC @jiachengxu ?

No, I am not using SSO RBAC feature for my setup

**Comment by jiachengxu** [comment]
I have updated the tests to use `Contains` instead of `Equal`: https://github.com/argoproj/argo-workflows/pull/13194/commits/bcbfef5117dfa1691fe12b90f1c9c83633d841fe

**Review by agilgur5 — APPROVED** [review]
> I have updated the tests to use `Contains` instead of `Equal`: [bcbfef5](https://github.com/argoproj/argo-workflows/commit/bcbfef5117dfa1691fe12b90f1c9c83633d841fe)

Yea that's what I usually do with these kinds of error message based tests.

LGTM, thanks for the update!

> i wonder if this helps with #13061 do u happen to use SSO RBAC @jiachengxu ?

@tooptoop4 we'll want someone to test with `:latest` once this is merged and the image is released. The `expr` changelog didn't mention anything that I thought would affect it, but 🤷

**Pull request merged**